types.cc

```diff
@@ -3333,24 +3158,176 @@ static bytes concat_fields(const std::vector<bytes>& fields, const std::vector<i
     return result;
 }
 
-bytes
-tuple_type_impl::from_string(sstring_view s) const {
-    std::vector<sstring_view> field_strings = split_field_strings(s);
-    if (field_strings.size() > size()) {
-        throw marshal_exception(format("Invalid tuple literal: too many elements. Type {} expects {:d} but got {:d}", as_cql3_type(), size(), field_strings.size()));
+namespace {
+struct from_string_visitor {
+    sstring_view s;
+    bytes operator()(const reversed_type_impl& r) { return r.underlying_type()->from_string(s); }
+    template <typename T> bytes operator()(const integer_type_impl<T>& t) { return t.decompose_value(t.parse_int(s)); }
+    bytes operator()(const string_type_impl&) {
+        return to_bytes(bytes_view(reinterpret_cast<const int8_t*>(s.begin()), s.size()));
+    }
+    bytes operator()(const bytes_type_impl&) { return from_hex(s); }
+    bytes operator()(const boolean_type_impl& t) {
+        sstring s_lower(s.begin(), s.end());
+        std::transform(s_lower.begin(), s_lower.end(), s_lower.begin(), ::tolower);
+        bool v;
+        if (s.empty() || s_lower == "false") {
+            v = false;
+        } else if (s_lower == "true") {
+            v = true;
+        } else {
+            throw marshal_exception(format("unable to make boolean from '{}'", s));
+        }
+        return t.make_value(v).serialize();
+    }
```

Comment:
Doesn't this (and similar) add a virtual call?  The original knows the type, but make_value() forgets the type and now we have a virtual call, unless the compiler is clever enough to inline everything.

Reply:
fixed